### DISCUSSION TITLE
Hashes with more than 32 byte (e.g. 64 or 128 byte) to avoid possible hash collisions

### DISCUSSION THREAD

**Discussion by mumpitzstuff** [ccache/ccache#951] [Ideas]
Ccache works for me most of the time but after some time of usage, i get sometimes corrupted executables. After clearing the cache and rebuilding the project everything works again. The problem is that a build can take up to 30min and to clear the cache and rebuild everything is not that useful.

It is near impossible to be sure what will happen in the background if the executable gets corrupted but i would say it has something to do with hash collisions (just a guess). Would it be possible to have a setting to allow longer hashes like 64 byte or 128 byte? This could reduce the chance of hash collisions drastically.

now: 05j3qb24p6ju75tn737qgm17bvnnek2R
new (64byte): 05j3qb24p6ju75tn737qgm17bvnnek2R0csp0nm7dhsk6jgvn84c5ulr56ah0dmM

Another option could be to add the original filename or something like that to the filename in the cache folder. All of that could be configurable.

now: 05j3qb24p6ju75tn737qgm17bvnnek2R
new: 05j3qb24p6ju75tn737qgm17bvnnek2R_original_filename_c

**Comment by jrosdahl** [comment]
> Ccache works for me most of the time but after some time of usage, i get sometimes corrupted executables. After clearing the cache and rebuilding the project everything works again.

Which ccache version are you using? This sounds like #935 which can affect cache entries created by ccache 4.4 and 4.4.1.

> It is near impossible to be sure what will happen in the background if the executable gets corrupted but i would say it has something to do with hash collisions (just a guess).

Ccache currently uses 160 bit hash digests (i.e. 20 bytes, not 32) from a cryptographic hash algorithm, and such algorithms are designed to have perfect uniformity. Say that you cache 1800 unique compilations each hour, i.e. 1 new cache entry per second on average (assuming each compilation produces one manifest and one result entry). To get a hash collision with the odds of one in a billion (10<sup>-9</sup>) you'll need to keep compiling unique source code files for about 1.7 trillion (1.7⋅10<sup>12</sup>) years. To have a 50% chance of a hash collision you'll need to keep going for 45 quadrillion (4.5⋅10<sup>16</sup>) years. (This reasoning assumes that you also keep all produced cache entries in your cache, so you'll need a large storage device. :slightly_smiling_face:)

So no, you don't need to worry about hash collisions. It's astronomically more likely that what you see is an ordinary bug.

That said, it would of course be possible to use a larger hash digest size, but I see little reason to do that.

**Comment by mumpitzstuff** [comment]
We had this issue all the time and long time before 4.4 was released. Therefore i dont think that it is related to #935.
Some times we can run our builds 2-3 weeks without any problems and then at some point we get a corrupted executable (build was successful but the executables will crash). And some time i just run a few builds in 2-3 days to get the same behaviour.
What i have done so far is the following. I made a backup of all objects at the time i got such an error. I disabled the ccache afterwards, rebuild the whole project and compared the objects. 99,9% of the objects are the more or less equal (hard to say because all the objects seems to include a date or something like that, which is always different). But in some objects (2-3 of 3.200) i have seen bigger differences than normal. It was not that easy to identify these files and i had no idea how to proceed after that. 

PS: I just copied a filename from my ccache folder to an editor and have seen a length of 32 chars. Therefore my assumption was that 32 byte hashes are used which seems to be wrong...

**Reply by jrosdahl** [reply]
> We had this issue all the time and long time before 4.4 was released. Therefore i dont think that it is related to #935.

OK. The maybe you're triggering [_The direct mode fails to pick up new header files in some rare scenario_](https://ccache.dev/manual/4.4.2.html#_caveats). Or some other bug, but it's not hash collisions.

> But in some objects (2-3 of 3.200) i have seen bigger differences than normal. It was not that easy to identify these files and i had no idea how to proceed after that.

The only way I know to debug things like this that is to enable `CCACHE_DEBUG` (and optionally `CCACHE_DEBUGDIR`), save debug info for all builds and compare the logs and input data between two compilations that should not have been the same but were the same with ccache.

> PS: I just copied a filename from my ccache folder to an editor and have seen a length of 32 chars. Therefore my assumption was that 32 byte hashes are used which seems to be wrong...

160 bits means 20 bytes. Those bytes have to be encoded as filesystem-friendly characters since not all bytes are valid in filenames. A common way to do that is to use [base16 encoding](https://en.wikipedia.org/wiki/Hexadecimal), which encodes 4 bits per character, so 40 hex characters for 160 bits. Ccache uses 4 base16 characters + 29 [base32hex](https://en.wikipedia.org/wiki/Base32#base32hex) characters to encode the 160 bits, and then adds a final type character (`R` or `M`).
